tests/test_core/test_component.py

```diff
@@ -0,0 +1,76 @@
+import functools
+
+from winter.core import Component
+from winter.core import is_component
+from winter.core.application import WinterApplication
+from winter.core.state_key import StateKey
+
+
+def test_is_component():
+    winter_app = WinterApplication()
+
+    @winter_app.component
+    class SimpleController:
+        pass
+
+    assert is_component(SimpleController)
+
+
+def test_methods():
+    winter_app = WinterApplication()
+
+    @winter_app.component
+    class SimpleComponent:
+
+        @winter_app.component_method
+        def simple_method(self):
+            return None
+
+    controller = Component(SimpleComponent)
+
+    assert len(controller.methods) == 1, SimpleComponent.simple_method
+    method = controller.methods[0]
+
+    assert method is SimpleComponent.simple_method
+
+
+def test_method_state():
+    winter_app = WinterApplication()
+
+    def route(param: str):
+        state_key = StateKey('path')
+        state_value = param
+        return functools.partial(winter_app.component_method, state_key=state_key, state_value=state_value)
+
+    @winter_app.component
+    class SimpleComponent:
+
+        @route('/url/')
+        def simple_method(self):
+            return None
+
+    state_key = StateKey('path')
+    assert SimpleComponent.simple_method.get_state(state_key) == '/url/'
+
+
+def test_method_state_many():
+
+    winter_app = WinterApplication()
+
+    def query_param(param: str):
+        state_key = StateKey('query', many=True)
+        state_value = param
+        return functools.partial(winter_app.component_method, state_key=state_key, state_value=state_value)
+
+    @winter_app.component
```

Comment:
It looks like a mix of auto-discovery and manual registration. That way `winter_app` will be spread across all the source code of application. Please consider splitting the mechanism into three parts:
* component annotation (does nothing except defining the class as an application component)
* manual component registration into an application (relies on the annotation)
* auto-discovery, which scans the source code for annotated classes and automatically registers them in an application

```
@winter.component
class SomeComponent:
    pass

winter_app = WinterApplication()

# Manually register components
winter_app.add_component(SomeComponent)

# OR use auto-discovery
winter_app.autodiscover()
```

Reply:
Done